modules/mod_ginger_rdf/controllers/controller_hydra.erl

```diff
@@ -152,3 +154,14 @@ page(QueryParams, PageNumber) when is_number(PageNumber) andalso PageNumber >= 1
 
 id(Context) ->
     m_rsc:rid(z_context:get_q(id, Context), Context).
+
+%% @doc If the resource has a query text, use that to search.
+%%      If not, fall back to the resource's outgoing haspart edges.
+-spec search_arguments(m_rsc:resource(), z:context()) -> proplists:proplist().
+search_arguments(Id, Context) ->
+    case m_rsc:p(Id, query, Context) of
+        <<>> ->
+            [{is_published, true}, {hassubject, [Id, haspart]}];
+        _QueryText ->
+            [{is_published, true}, {query_id, Id}]
+    end.
```

Comment:
For collections containing both query and haspart links, this would result in omitting haspart data - is this a compromise/as intended?

Reply:
Definitely a compromise that could be improved upon. However, search pagination will be incorrect if we simply add the two sets of resources together (those from the dynamic query and the haspart objects).

It would be more elegant to combine the search criteria to say: give me all resources `({query_id, Id} OR {hassubject, [Id, haspart]}) AND is_published`. However, I’ve found no way to express this in a Zotonic query.

Reply:
Yes indeed, agreeed.